### DISCUSSION TITLE
query parameters and headers for 3D tiles loader

### DISCUSSION THREAD

**Discussion by ibgreen** [visgl/loaders.gl#2470] [General]
@jo-chemla In response to your question in the closed PR https://github.com/visgl/loaders.gl/pull/2252#issuecomment-1546748427

There has been a fair amount of hacking by multiple folks over time, so it can get confusing and probably benefits from some cleanup and documentation. 

The deck.gl example at https://deck.gl/examples/google-3d-tiles (source code at https://github.com/visgl/deck.gl/blob/8.9-release/examples/website/google-3d-tiles/app.jsx#L53) uses loadOptions on the layer to set the X-GOOG-API-KEY. This results in fetch options being passed in the initial tileset load.

The PR in 3.3.5  you reference was intended to allow you to simply pass query parameters as part of the original URL (i.e. no need to pass them in through undocumented tileset properties).

**Comment by arthurschiller** [comment]
Hi, could @ibgreen or @felixpalmer please share an example on how this would work?
I want to use the new google maps 3D tiles API with a custom renderer (https://developers.google.com/maps/documentation/tile/create-renderer). 

So I'm first fetching a root tileset to retrieve a session. Now I have a URL like: `https://tile.googleapis.com/v1/3dtiles/datasets/CgA/files/someFile.json?session=mySession&key=myKey`

which I want to feed into the loader:
`const tilesetJson = await load(url, Tiles3DLoader);
const tileset = new Tileset3D(tilesetJson);`

This works but then it tries to resolve the glbs and fails, because the query parameters apparently have been removed:
`https://tile.googleapis.com/v1/3dtiles/datasets/CgA/files/someTile.glb`

`[{
  "error": {
    "code": 403,
    "message": "The request is missing a valid API key.",
    "errors": [
      {
        "message": "The request is missing a valid API key.",
        "domain": "global",
        "reason": "forbidden"
      }
    ],
    "status": "PERMISSION_DENIED"
  }
}
]`

It should be:
`https://tile.googleapis.com/v1/3dtiles/datasets/CgA/files/someTile.glb?session=mySession&key=myKey`

What am I missing? Sorry if there is something obvious documented somewhere, I just got started working with the library.

I'm using the 3.4.0 tags:
`@loaders.gl/core@3.4.0-alpha.4 and @loaders.gl/3d-tiles@3.4.0-alpha.4`

Thanks a bunch!

**Reply by ibgreen** [reply]
Per my understanding, the `session` parameter should not be supplied by the application. It is generated internally by google tiles API (returned in URLs in the tileset.json files). Instead, the parameter that should be supplied to the top tile load is the api key.

We did add code to preserve the session parameter before the launch but perhaps we did not get the support for key parameter in top level URL working.

The only working example I am aware of is the deck.gl example. 

@felixpalmer has a more detailed understanding.

**Reply by arthurschiller** [reply]
Thanks for the swift reply! Yes the session is already included in the retuned response URI, something like: `/v1/3dtiles/datasets/CgA/files/UlRPVEYuYnVs.json?session=CIqhrPOFvdHSYg` from that and the base URL I'm creating something like: `https://tile.googleapis.com/v1/3dtiles/datasets/CgA/files/UlRPVEYuYnVs.json?session=CIqhrPOFvdHSYg&key=myKey`and feed that into the loader.

Okay will have a look into the deck.gl example. Tile3DLayer also uses Tiles3DLoader and Tileset3D under the hood?

**Reply by arthurschiller** [reply]
From what I'm seeing I have the same setup as outlined here:

<img width="858" alt="Screenshot 2023-05-15 at 14 21 12" src="https://github.com/visgl/loaders.gl/assets/3637723/62edeb1f-8e71-4e5d-a0e8-2e28e878d86b">

But if I check `tileset.queryParams` it is empty.

**Reply by ibgreen** [reply]
OK, since that is a passing test case, more info might be necessary, e.g: 

- are you actually creating a Tileset3D instance this way and getting different results?
- Or are you using deck.gl Tile3DLayer with the (reasonable) assumption that this will work the same way?
- Or some other approach?

**Reply by jo-chemla** [reply]
Thanks for your quick replies. On my end, I'm using the example from the loaders.gl docs [here](https://loaders.gl/docs/modules/3d-tiles/api-reference/tiles-3d-loader#usage), with no deck.gl apart from the `WebMercatorViewport` constructor. 

 - Root tileset is fetched correctly (by providing either the url params or the header key-value pair), at that url: `https://tile.googleapis.com/v1/3dtiles/root.json`
 - But the subsequent tile (the first children of root) is fetched from the url provided by the tileset root `root.children[0].children[0].content.uri`, without adding back urlParams or header keys. `/v1/3dtiles/datasets/CgA/files/{TILE_ID}.json?session={SESSION}`

```js
const tilesetJson = await load(tilesetUrl, Tiles3DLoader, {
  "3d-tiles": {
    isTileset: true,
  },
  fetch: {
    headers: {
      "X-GOOG-API-KEY": GOOGLE_MAPS_API_KEY,
    },
    queryParams: { test: 100000000 },
    queryString: "a=123&b=abc",
  },
});
const tileset3d = new Tileset3D(tilesetJson, {
  onTileLoad: (tile) => console.log(tile),
});

const viewport = new WebMercatorViewport({
  latitude, longitude, zoom
});
tileset3d.update(viewport);
const visibleTiles = tileset3d.tiles.filter((tile) => tile.selected);
```

**Reply by ibgreen** [reply]
There has been several people making quick fixes in this part of loaders.gl, but I can say

- allowed `fetch` options should only be options accepted by fetch: https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters.

I see in the code that the tilesetJson allows for a `queryString` to be specified. I am not convinced that is the long term solution, but for now perhaps you can slap it in there before creating the Tileset3D.

Currently a lot of work or loaders.gl 4.0 revolves around on strongly typing loader options and return values, which will hopefully take some of the guess work out of where parameters can be supplied.

**Reply by arthurschiller** [reply]
> OK, since that is a passing test case, more info might be necessary, e.g:
> 
> * are you actually creating a Tileset3D instance this way and getting different results?
> * Or are you using deck.gl Tile3DLayer with the (reasonable) assumption that this will work the same way?
> * Or some other approach?

My current playground setup basically looks like this: https://gist.github.com/arthurschiller/5b2b2a1f0ba087f7f155416ed18b3e8d

Basically option A.

**Comment by felixpalmer** [comment]
@arthurschiller the way we have it working via deck (which passes the parameters through to loaders.gl) is to only provide the API key as a header, this way it is persistent:

https://github.com/visgl/deck.gl/blob/master/examples/website/google-3d-tiles/app.jsx#L53-L55

This [PR](https://github.com/visgl/loaders.gl/pull/2456/files#diff-a648053d407c39225dddd6336b07c713660253c4eec91e8d3eb8545193bd8dfd) added automatic support for the session parameter

**Reply by arthurschiller** [reply]
Thank you @felixpalmer!
loaders.gl tag 3.4.0-alpha.4 is the right one to use for now?
I've gotten a step further but now I'm getting: `3D tile failed to load: undefined 3DTileLoader: unknown type {"as
_` after setting the viewport.

Setup is:

	import { load } from '@loaders.gl/core';
	import { Tiles3DLoader } from '@loaders.gl/3d-tiles';
	import { Tileset3D } from '@loaders.gl/tiles';
	import { WebMercatorViewport } from '@deck.gl/core';

	const tilesetUrl = 'https://tile.googleapis.com/v1/3dtiles/root.json';
	const loadOptions = {
		fetch: {
			headers: { 'X-GOOG-API-KEY': apiKey }
		},
		'3d-tiles': { 
			isTileset: true,
			loadGLTF: false
		}
	};

	const tilesetJson = await load(
		tilesetUrl, 
		Tiles3DLoader,
		loadOptions
	);
	console.log('tilesetJson: ', tilesetJson);

	const tileset3d = new Tileset3D(
		tilesetJson,
		{
			loadOptions,
			onTileLoad: (tile) => {
				console.log('New tile loaded.')
				console.log(tile)
			},
			onTileError: (tile, error) => {
				console.log('Tile Error: ', error)
			}
		}
	);
	console.log('tileset3d: ', tileset3d);
	const viewport = new WebMercatorViewport({
		width: 600,
		height: 600,
		longitude: 0,
		latitude: 0,
		zoom: 10
	});

	// set initial viewport
	tileset3d.update(viewport);

**Reply by arthurschiller** [reply]
Seems like this error only happens with v3.4.0-alpha.4 but v3.3.5 works. 👍
I also gotta figure out now why it's not traversing but should be solvable. Thanks a bunch!

**Reply by felixpalmer** [reply]
Please report back once you have a working solution so we can update in the docs

**Comment by felixpalmer** [comment]
@jo-chemla when constructing the `Tileset3D` you need to include the loadOptions:

```
const tilesetUrl = 'https://tile.googleapis.com/v1/3dtiles/root.json';
const loadOptions = {fetch: {headers: {'X-GOOG-API-KEY': GOOGLE_MAPS_API_KEY}}};

export async function run() {
  const tilesetJson = await load(tilesetUrl, Tiles3DLoader, loadOptions);
  const tileset3d = new Tileset3D(tilesetJson, {loadOptions});
  const viewport = new WebMercatorViewport({
    width: 600,
    height: 600,
    longitude: 0,
    latitude: 0,
    zoom: 10
  });
  tileset3d.update(viewport);
}
```

**Reply by jo-chemla** [reply]
Alright, thanks for that comment - I indeed passed the `loadOptions` only to the `tilesetJson = await load(...)` instead of also to the `Tileset3d` constructor. 
Now I'll need to understand why the library is only loading a single tile and not traversing the tree to fetch all tiles within the `WebMercatorViewport` I've used in the `tileset3d.update(viewport)` call, but that's another story! Probably has to do with the `geometricError` parameter I need to tweak. Thanks again for your help!

**Comment by taylorwebber** [comment]
@arthurschiller @felixpalmer was there ever any resolution on this issue?

I'm running into the same issues stated above where the session is not being added as a query parameter and thus throwing 400 bad request errors when attempting to fetch the tiles after updating the tileset with the viewport.
